### DISCUSSION TITLE
up unwind for aarch64 by creating a new context not working.

### DISCUSSION THREAD

**Discussion by shenmufeng** [libunwind/libunwind#640] [Q&A]
I have been using unwind a lot in x86_64, and it works fine. Now I want to use it on my aarch64 but faced a problem.

My code is basically to making up unwinding:

1. In my code, when up unwind fail(for internal reason), it will try to step forward by using rbp  like this(this is only for x86_64, I don't know how to do this in aarch64):
    frame->rsp = frame->rbp+2;
    frame->rip = *(frame->rbp+1);
    frame->rbp = *frame->rbp;

2. Then I will continue using libunwind to start from new point like this(offset for aarch64 from offset.h):

#define LINUX_UC_MCONTEXT_OFF   0xb0
#define UC_REG_BP   (LINUX_UC_MCONTEXT_OFF + 0x0f0) //X29 -> FP(in aarch64) 416 
#define UC_REG_SP   (LINUX_UC_MCONTEXT_OFF + 0x100) //X31 -> SP     432 
#define UC_REG_IP   (LINUX_UC_MCONTEXT_OFF + 0x0f8) //X30 -> LR(in aarch64) 424
#define UNW_REG_BP UNW_AARCH64_X29

             unw_context_t uc;
             unw_ret = unw_getcontext(&uc);

             char *p_uc = (char*)&uc;
            *((void **)((char*)p_uc + UC_REG_BP)) = frame->rbp;
            *((void **)((char*)p_uc + UC_REG_SP)) =  frame->rsp;
            *((void **)((char*)p_uc + UC_REG_IP)) = frame->rip;     
           
              unw_ret = unw_init_local(cursor, &uc);
          
              unw_ret = unw_step(cursor);

3. for aarch64, unw_step is not crash, but just believe it reaches end now. For x86_64, it works just fine.

4. So my question is, how could I do corresponding aarch64 for the same purpose? I don't really understand how to step up unwind for those lr,fp,pc registers. Thanks!!!

**Comment by bregma** [comment]
One of the problems with doing this for aarch64 is that you can't always easily find the saved LR and FP values in a stack frame because the stack frame size varies depending on the number of automatic variables in the procedure and some compilers store the LR/FP pair at the _bottom_ of the stack frame. This is why the DWARF program (from the FDE)  needs to be run during unwinding to determine the original values of those registers so they can be restored.
